Result/Result.swift

```diff
@@ -179,6 +179,15 @@ public func ?? <T, Error> (left: Result<T, Error>, @autoclosure right: () -> Res
 	return left.recoverWith(right())
 }
 
+// MARK: - Derive result from failable closure
+
+public func materialize<T>(f: () throws -> T) -> Result<T, ErrorType> {
```

Comment:
> It's better to keep type info, but how is E inferred?

I haven't experimented with this case, but generally the type-checker can infer from the surrounding code without explicit specification at the call-site.

> Currently, we can't know what kind of error () throws -> T throws.

I suspect there are a lot of cases where you "know" the error, it just can't be documented in the throws clause. And in the cases where you don't, you can use `ErrorType` directly.